File: Week_04/binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (!root)  return res;
        // 使用两个 vector 分别保存当前层和下一层的节点
        vector<TreeNode *> cur, next;
        // 保存当前层的数据
        vector<int> vals;

        cur.push_back(root);
        while (!cur.empty() || !next.empty()) {
            // 遍历当前层节点
            for (int i = 0; i < cur.size(); i++) {
                TreeNode *node = cur[i];
                vals.push_back(node->val);
                if (node->left) next.push_back(node->left);
                if (node->right) next.push_back(node->right);
            }
            // 先清空当前层节点，同时指向下一层节点，最后下一层节点清空
            cur.clear();
            cur = next;
            next.clear();
            // 将当前层的数据保存，然后清空vector
            res.push_back(vals);
            vals.clear();
        }
        return res;
    }
};

File: Week_04/binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if root == None:
            return []

        vals = []
        cur_level_vals = []
        cur_level_nodes = [root]
        next_level_nodes = []
        while True:
            while cur_level_nodes:
                node = cur_level_nodes.pop(0)
                if node.left:
                    next_level_nodes.append(node.left)
                if node.right:
                    next_level_nodes.append(node.right)
                cur_level_vals.append(node.val)
            vals.append(cur_level_vals)
            cur_level_vals = []
            if next_level_nodes:
                cur_level_nodes = next_level_nodes
                next_level_nodes = []
            else:
                break
        return vals
